xds/internal/resolver/xds_resolver_test.go

```diff
@@ -858,60 +1036,115 @@ func (s) TestXDSResolverRemovedResource(t *testing.T) {
 	}
 }
 
-func (s) TestXDSResolverWRR(t *testing.T) {
-	xdsR, xdsC, tcc, cancel := testSetup(t, setupOpts{target: target})
-	defer xdsR.Close()
-	defer cancel()
-
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer cancel()
-	waitForWatchListener(ctx, t, xdsC, targetStr)
-	xdsC.InvokeWatchListenerCallback(xdsresource.ListenerUpdate{RouteConfigName: routeStr, HTTPFilters: routerFilterList}, nil)
-	waitForWatchRouteConfig(ctx, t, xdsC, routeStr)
-
+// TestResolverWRR tests the case where the route configuration returned by the
+// management server contains a set of weighted clusters. The test performs a
+// bunch of RPCs using the cluster specifier returned by the resolver, and
+// verifies the cluster distribution.
+func (s) TestResolverWRR(t *testing.T) {
 	defer func(oldNewWRR func() wrr.WRR) { newWRR = oldNewWRR }(newWRR)
 	newWRR = testutils.NewTestWRR
 
-	// Invoke the watchAPI callback with a good service update and wait for the
-	// UpdateState method to be called on the ClientConn.
-	xdsC.InvokeWatchRouteConfigCallback("", xdsresource.RouteConfigUpdate{
-		VirtualHosts: []*xdsresource.VirtualHost{
-			{
-				Domains: []string{targetStr},
-				Routes: []*xdsresource.Route{{Prefix: newStringP(""), WeightedClusters: map[string]xdsresource.WeightedCluster{
-					"A": {Weight: 5},
-					"B": {Weight: 10},
-				}}},
-			},
-		},
-	}, nil)
+	mgmtServer, err := e2e.StartManagementServer(e2e.ManagementServerOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer mgmtServer.Stop()
 
+	// Create a bootstrap configuration specifying the above management server.
+	nodeID := uuid.New().String()
+	cleanup, err := xdsbootstrap.CreateFile(xdsbootstrap.Options{
+		NodeID:    nodeID,
+		ServerURI: mgmtServer.Address,
+		Version:   xdsbootstrap.TransportV3,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cleanup()
+
+	// Build an xDS resolver that uses the above bootstrap configuration
+	// Creating the xDS resolver should result in creation of the xDS client.
+	builder := resolver.Get(xdsScheme)
+	if builder == nil {
+		t.Fatalf("resolver.Get(%v) returned nil", xdsScheme)
+	}
+	const serviceName = "my-service-client-side-xds"
+	ldsName := serviceName
+	rdsName := "route-" + serviceName
+	u, err := url.Parse("xds:///" + serviceName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	tcc := newTestClientConn()
+	r, err := builder.Build(resolver.Target{URL: *u}, tcc, resolver.BuildOptions{})
+	if err != nil {
+		t.Fatalf("builder.Build(%v) returned err: %v", target, err)
+	}
+	defer r.Close()
+
+	// Configure the management server with a good listener resource and a
+	// route configuration resource.
+	resources := e2e.UpdateOptions{
```

Comment:
Should we move this also to a helper function just to nix some duplicate code here and 2 more times above?

Reply:
I understand that the resources occupy a lot of lines here, and in other tests. But it is hard to move them to helpers because each test might build the resource in a slightly different way. Once all tests are modified, I will make a pass to see if I can refactor out some resources to reduce boiler plate.

Reply:
gotcha